proof/invariant-abstract/RealTime_AI.thy

```diff
@@ -1054,12 +1054,19 @@ crunches
   (wp_del: get_sc_active_wp get_sc_refill_ready_wp get_sc_refill_sufficient_wp
            get_sc_released_wp refill_full_wp)
 
+(* FIXME RT Michael: move to lib *)
+lemma whileLoop_wp':
+  "(\<And>r. \<lbrace> \<lambda>s. I r s \<and> C r s \<rbrace> B r \<lbrace> I \<rbrace>) \<Longrightarrow>
+  \<lbrace> I r \<rbrace> whileLoop C B r \<lbrace> I \<rbrace>"
+  apply (fastforce intro: whileLoop_wp)
+  done
+
```

Comment:
Since `whileLoop_wp'` is used in so many crunches proofs, is it sensible to add it to `wp` or `crunch_wps` instead of adding it in so many places? In the cases were you don't want to use it, which would be few, you could `supply whileLoop_wp' [wp del]`. What do you think?

Reply:
I'm not even sure how to add it to `crunch_wps`. @corlewis can you let me know what you think of this, and how to do it if it's a good idea?

Reply:
Hmm, at the moment the only way to add things to `crunch_wps` is by manually adding it when it is first created, which is currently in Bits_AI. This would require you following through with the move to lib.

Separately, I definitely don't think it should go in`wp` but it might be ok in `crunch_wps`. In general we really only want safe rules in `crunch_wps` and especially in `wp`. I haven't looked at this PR enough yet to really tell how safe this rule is, but my guess would have been that this can only be used in cases where the body doesn't really interact with the condition. More complicated cases will need `whileLoop_wp` and I think it's too confusing to expect people to know to delete `whileLoop_wp'` first.

The reason that I think it might be ok for `crunch_wps` is that you're probably only going to be able to crunch functions with loops when it's for a property that's independent of the loop. Anything else seems like it would already be too complicated for crunch without some specific helper lemmas. @lsf37 what do you think?

Reply:
Yes, that makes sense to me. We probably don't want loop rules in general `wp`, but for invariants in `crunch_wps` it would make sense.